keps/sig-cluster-lifecycle/kubeadm/20190925-component-configs.md

```diff
@@ -118,78 +119,79 @@ Currently, kubeadm vendors in pieces of Go code from the components to be able t
+### Stop using component config internal types
 
-##### User supplied Kubernetes Core Component Config management
+As kubeadm will no longer default, validate or migrate component configs inside its code base, the need to use component config internal types is no longer needed. In that case the usage of internal types can be removed. This would also allow us to reduce the number of imported packages and to drop one of the last dependencies on the kubernetes/kubernetes code base.  
+The switch is going to be performed by replacing each internal type with a corresponding versioned type. This, by itself, is not going to have any impact on users.
 
-As user changes need to be maintained, user supplied configs are stored config maps.  
-However, in those cases kubeadm is not responsible for config migration upon upgrade. If the new component version is unable to consume the user supplied config, kubeadm is going to bail out with an error message, prompting the user to perform a manual config migration beforehand. This is a behavior similar to that for addon configs.
+### “kubeadm upgrade plan” changes
 
-##### Allow Kustomize use over core Kubernetes component configs
+As supported component configs are expected to increase in number in the future and some of them might require manual user migration prior to upgrade, an ahead of time notice about what actions are required is a nice thing to have.  
+Hence, the `kubeadm upgrade plan` command is going to provide information on the component configs. In particular, if they are managed by kubeadm or are user supplied, and if they need manual upgrade or not.
 
-Kubeadm shall extend its current Kustomize support to include patching of core Kubernetes component configs. Currently, this will include only kubelet’s component config. However, in the future this might extend to the API Server, Scheduler and Controller Manager as well.  
-Most notably, this feature is not going to be applied to kube-proxy’s config by kubeadm as this is an addon component config and the responsibility for this is going to be with the corresponding addon operator.
+In addition to that, `kubeadm upgrade plan` is now also going to support a new flag `--component-config-upgrades` that is going to supply a file with one or more YAML documents in it that overwrite component configs that are too old and cannot be loaded from the cluster.  
+Only the configs that need upgrading would be loaded from the file supplied with `--component-config-upgrades`. Component configs that don't need version upgrades or kubeadm configuration types will be ignored.  
+The loaded configs, that were supplied with `--component-config-upgrades` would be taken into account when printing the additional component config information.
```

Comment:
this is the only new topic, i guess.
i think it's fine other that adding another flag that will be difficult to remove later.
wondering if we should have it prefixed as `--experimental` so that we can have easier deprecation.

my idea was that maybe we should introduce an UpgradeConfiguration that can host:
- potential custom image changes, where the user wants to upgrade their custom etcd to this new custom etcd version.
- component config upgrade.

Reply:
> this is the only new topic, i guess.

Plus the new alpha commands to make upgrade UX easier.

> i think it's fine other that adding another flag that will be difficult to remove later.
> wondering if we should have it prefixed as `--experimental` so that we can have easier deprecation.

The current flag name is the first thing that came in my mind. Sure, we should prepend `experimental` while this whole UX is still alpha. However...
 
> my idea was that maybe we should introduce an UpgradeConfiguration that can host:
> 
> * potential custom image changes, where the user wants to upgrade their custom etcd to this new custom etcd version.
> * component config upgrade.

... I don't think that embedding component configs in another kubeadm config object is handy or appropriate even. This would result into a massive inconsistency with the init UX where component configs are specified as separate YAML documents (which is the same way kubectl and the rest of the Kubernetes community operates).

I have other reservations towards `UpgradeConfiguration` too, but they are off-topic WRT this PR.

Hence, I don't see any other way than specify a file with YAML documents to `kubeadm upgrade apply`. I didn't want to reuse `--config` though, because that would make our life harder once we try to remove that capability in the future. Plus, it makes a better distinction between the semantic of "using everything there" (that `--config` has) vs "use only what is necessary" (that `--component-config-upgrades` has).

Reply:
but this is also an inconsistency with `init`, which allows passing component config in a mulit-doc.
i'm leaning towards `--experimental-component-config-upgrade` (singular) and maybe we can make `--config` only used for that.

Reply:
Actually,
```
In addition to that, `kubeadm upgrade plan` is now also going to support a new flag `--component-config-upgrades` that is going to supply a file with one or more YAML documents in it that overwrite component configs that are too old and cannot be loaded from the cluster.
```

Hence you can specify multiple YAMLs in the same file, but only the needed ones will be used (unneeded stuff will be ignored and this includes kubeadm configs).

Reply:
marking as resolved.